From 8778a83d1a403a614496b9e7904debd808ca6e73 Mon Sep 17 00:00:00 2001
From: Terry Zhou <bjzhou@marvell.com>
Date: Thu, 28 Dec 2017 14:20:22 +0800
Subject: [PATCH 1256/1345] fix: mvebu: rtc: write twice to ensure RTC time
 setting

commit  3f978bf4c5c50528b3907b425fcc6370e9d93a2a from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

When rapidly writing to RTC register and reading it back,
sometimes the new time value didn't take effect and keep
with the existing value. According to the datasheet, the
RTC time register should be written twice to guarantee
the write is performed correctly.

Change-Id: Iab96cc48a0b1fe7fd027aad6b164b46992ce0e70
Signed-off-by: Terry Zhou <bjzhou@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/50421
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
Tested-by: Wilson Ding <dingwei@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/rtc/rtc-armada38x.c |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/rtc/rtc-armada38x.c b/drivers/rtc/rtc-armada38x.c
index eae5723..10b48a3 100644
--- a/drivers/rtc/rtc-armada38x.c
+++ b/drivers/rtc/rtc-armada38x.c
@@ -99,6 +99,8 @@ struct armada38x_rtc_data {
  * may fail. As a workaround, before writing to RTC
  * register, issue a dummy write of 0x0 twice to RTC Status
  * register.
+ * According to the datasheet, the RTC time register should be written
+ * twice to guarantee the write is performed correctly
  */
 
 static void rtc_delayed_write(u32 val, struct armada38x_rtc *rtc, int offset)
@@ -107,6 +109,8 @@ static void rtc_delayed_write(u32 val, struct armada38x_rtc *rtc, int offset)
 	writel(0, rtc->regs + RTC_STATUS);
 	writel(val, rtc->regs + offset);
 	udelay(5);
+	writel(val, rtc->regs + offset);
+	udelay(5);
 }
 
 /* Update RTC-MBUS bridge timing parameters */
-- 
1.7.9.5

